### PR TITLE
Enhancement: Use matrix to reduce duplication

### PR DESCRIPTION
**Pull request by localheinz**
This pull request

- [x] uses build matrixes to reduce the duplication in GitHub Actions workflow definitions

### CODE REVIEW

File: .github/workflows/v2.yaml [hunk 1]
```diff
@@ -69,228 +76,52 @@ jobs:
...
       # Build full image: binary with runtime
       - name: Build full image
+        working-directory: ${{ matrix.version }}
         run: |
           docker build \
           --pull \
           --no-cache \
           --target binary-with-runtime \
           --tag composer/composer:latest \
           --tag composer/composer:2 \
```
Comment (localheinz) [inline .github/workflows/v2.yaml:109]: Unless I am mistaken, the only difference between the proposed `build-previous` and `build-latest` version is that we use a tag `2` in addition to `2.x`. Is that right?

File: .github/workflows/v2.yaml [hunk 2]
```diff
@@ -11,36 +11,43 @@ env:
...
 jobs:
-  build22:
-    name: Build v2.2
+  build-previous:
+    name: Build previous versions
     runs-on: ubuntu-latest
     timeout-minutes: 20
-    defaults:
-      run:
-        working-directory: '2.2'
```
Comment (localheinz) [inline .github/workflows/v2.yaml:20]: Can't use expressions here, that's why I pushed these configurations down.

See https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#defaultsrun.

File: .github/workflows/v2.yaml (hunk 1 above)
Comment (alcohol) [reply .github/workflows/v2.yaml:109]: `latest` and `major` tag version

Comment (alcohol) [reply .github/workflows/v2.yaml:109]: Oh, and `2.2` is also tagged as `lts` to make it a bit more complicated

**Comment by alcohol** [comment]
I like where this is going. It doesn't take into account RC releases either currently, but those were also awkward in the past with the current setup. I was only able to review it on my phone right now so might have missed some minor details, but if you feel confident about it I'll happily merge it.

**Comment by localheinz** [comment]
@alcohol 

I feel confident, but I'm sure we'll find out if it works as expected!

**Comment by localheinz** [comment]
Thank you, @alcohol - I will be around for fixes!

**Comment by alcohol** [comment]
Thank you for taking the time to tidy this up 😁

**Comment by localheinz** [comment]
@alcohol 

Looks like I broke something, see https://github.com/composer/docker/actions/runs/6064410999:

CleanShot 2023-09-03 at 14 19 56@2x

**Comment by alcohol** [comment]
> @alcohol 
> 
> Looks like I broke something, see https://github.com/composer/docker/actions/runs/6064410999:
> 
> CleanShot 2023-09-03 at 14 19 56@2x
> 

Yeah you also need to change directory in the login/push steps since they grep for the specific version in the `Dockerfile`

**Comment by localheinz** [comment]
@alcohol 

Give me a moment!

**Comment by localheinz** [comment]
@alcohol 

Looks better now, thank you for pointing the `working-directory` out!

**Pull request merged**